Feat/#114 Sumary 그래프들의 UI 및 기능 구현

soralee2821 opened the pull request:
## Issues
- Issue number #114 

## Tasks Done 
- [x] 전체 풀업 운동의 총 운동 횟수 그래프 UI 구현
- [x] 각 풀업 운동에 대한 월별 운동 횟수 그래프 UI 구현
- [x] 운동 시간 대별 완료된 플랜 갯수 그래프 UI 구현
- [x] API 모킹
  - [x] 전체 풀업 운동의 총 운동 횟수 그래프 API
  - [x] 각 풀업 운동에 대한 월별 운동 횟수 그래프 API
  - [x] 운동 시간 대별 완료된 플랜 갯수 그래프 API

<br>

## Description
- React Chart 라이브러리인 [**Recharts**](https://recharts.org/en-US)를 사용하여 `Summary` 페이지에서 사용하는 그래프를 구현했습니다.
- **Recharts** 라이브러리를 선택한 이유
  - 공식 문서에 각 그래프에 대한 API와 예시가 잘 설명되어 있어서 사용하기 쉬웠습니다.
  - 그래프의 디자인이 간단하면서 데이터가 잘 나타나 있어서 디자이너가 좋아했습니다.

<br>

### TotalCountByAllWorkoutsChart (전체 풀업 운동의 총 운동 횟수 그래프)
- 전체 풀업 운동의 총 운동 횟수 그래프를 **Radar Chart**와 **Poloar Grid**를 사용하여 나타냈습니다.
- 각 풀업 운동의 총 횟수를 비교하여 최댓값을 가진 풀업 운동 이름을 문구로 명시했습니다.

<br>

### MonthlyTotalCountByEachWorkoutChart (각 풀업 운동에 대한 월별 운동 횟수 그래프)
- 이번 달을 포함한 지난 6달 동안 각 풀업 운동의 월별 횟수 그래프를 **Line Chart**와 **Cartesian Grid**를 사용하여 나타냈습니다.
- **antd** 라이브러리의 Dropdown 컴포넌트를 사용하여 `MonthDropdown`을 구현했습니다. 
  - `MonthDropdown`을 사용하여 풀업 운동을 선택할 수 있습니다.
  - 선택한 풀업 운동 이름을 **workoutName** 상태로 저장하여 사용했습니다.
  - **antd** 라이브러리의 **ConfigProvider** 컴포넌트를 사용하여 `MonthDropdown`의 디자인이 `ProfileImageDrowdown`과 동일하도록 설정했습니다.
- **workoutName** 상태에 저장된 풀업 운동에 대한 그래프를 나타내도록 구현했습니다.
  - `COLOR_BY_WORKOUT` 상수를 사용하여 선 그래프의 색상이 `Roadmap` 페이지의 해당 풀업 운동 색상과 동일하게 설정했습니다.
- 6달 동안 해당 풀업 운동의 총 운동 횟수를 비교하여 최댓값을 가진 달 이름을 문구로 명시했습니다.
  - `FULL_MONTH_NAME_BY_ABBREVIATION` 상수를 사용하여 데이터의 축약된 이름이 아니라 전체 이름을 명시하도록 했습니다.
 
<br>

### CompletedPlanCountByTimeChart (운동 시간대 별 완료된 플랜 갯수 그래프)
- 이번 달과 지난 달의 운동 시간대(morning, afternoon, evening)별 완료된 플랜 갯수 그래프를 **Bar Chart**와 **Cartesian Grid**를 사용하여 나타냈습니다.
- 이번 달의 운동 시간대 별 총 횟수를 비교하여 최댓값을 가진 운동 시간을 문구로 명시했습니다.

<br>

### 가입일로부터 오늘까지의 일수 표시
- **dayjs**의 **diff** 함수를 사용하여 오늘과 가입일의 차이를 계산하는 유틸 함수 `getDaysSpecificDate`를 구현했습니다.
  - 가입일뿐만 아니라 다른 날을 기준으로 오늘까지의 일수를 계산할 수 있다고 생각하여 함수 이름을 범용적으로 `getDaysSpecificDate`로 정했습니다.
- `useDataAtom`의 **joinDate** 값과 `getDaysSpecificDate` 함수를 사용하여 가입일로부터 오늘까지의 일 수를 계산하여 문구로 명시했습니다.

<br>

### 로그인 상태가 아닐 때 로그인 페이지로 리다이렉트
- `loginStateAtom`이 false일 때, **useNavigate** 훅을 사용하여 `Login` 페이지로 이동하도록 구현했습니다.

<br>

### API 모킹
- FE-BE 인터페이스 정의서의 [Summary API 명세](https://www.notion.so/dev-millie/Dashboard-Summary-API-eb248f7875ac451bafb72bbc8ff20227)를 참고하여 **MSW**를 통해 API를 모킹했습니다.
  - API 명세에 따라 **apiPath**, **queryKey**를 추가하고 해당 API를 호출하는 함수를 구현했습니다.
  - 해당 API를 모킹하기 위해 **mock data**와 **handler**를 추가했습니다.
  - 각 그래프 컴포넌트에서 상수 대신 해당 API를 호출하도록 수정했습니다.

<br>

### 참고 사항
- `ProfileImageDropdown`과 `MonthDropdown` 컴포넌트의 `customColorToken`은 같지만, 컴포넌트 내 자식 요소들이 다르기 때문에 두 컴포넌트를 따로 선언했습니다.
  - 두 컴포넌트 이외에도 모든 컴포넌트의 `cutomColorToken`을 동일하게 맞출 생각이라면 `lib/antd` 폴더 내에서 `customColorToken/dropdown` 파일을 만들어서 관리하면 좋을 것 같다는 생각이 들었습니다.
- 추후에 로딩 페이지 디자인이 완성되면 그래프의 데이터를 모두 받을 때까지 로딩 페이지를 보여주면 좋을 것 같다고 생각합니다.

<br>

## Visuals
![Summary](https://github.com/Pullanner/pullanner-web/assets/70058081/a06f83b0-f5ac-4d7b-8447-2dc3633d8af8)

soralee2821 committed:
[#114] Chore: install recharts package

soralee2821 committed:
[#114] Feat: add TotalCountByAllWorkoutsChart UI

- Recharts 라이브러리의 Radar Chart와 PolarGrid를 사용함
- 모든 풀업 운동의 총 횟수를 비교하여 최댓값을 가진 풀업 운동을 문구에 명시함

soralee2821 committed:
[#114] Feat: add MonthDropdown component

- Antd 라이브러리의 Dropdown을 사용하여 MonthlyTotalCountByEachWorkoutChart 그래프를 월별로 보여주기 위해 필요한 드롭다운을 구현함
- 드롭다운의 메뉴 아이템을 클릭하면 부모 컴포넌트 MonthlyTotalCountByEachWorkoutChart의 workoutName 상태가 변경됨

soralee2821 committed:
[#114] Feat: add constants to use Summary Page

- COLOR_BY_WORKOUT : 풀업 운동 이름 키와 색상 값을 가진 객체
- FULL_MONTH_NAME_BY_ABBREVIATION : 월 축약 이름 키와 축약하지 않은 이름 값을 가진 객체

soralee2821 committed:
[#114] Feat: add MonthlyTotalCountByEachWorkoutChart UI

- Recharts 라이브러리의 Line Chart와 Cartesian Grid를 사용하여 workoutName 상태 값에 해당하는 월별 운동 그래프를 구현함
- MonthDropdown을 사용하여 workoutName을 선택함
- COLOR_BY_WORKOUT을 통해 해당 workoutName에 대한 색상을 사용함
- 6달의 총 횟수를 비교하여 최댓값을 가진 월을 문구에 명시함
- FULL_MONTH_NAME_BY_ABBREVIATION을 통해 축약된 달 이름을 축약되지 않은 이름으로 변경함

soralee2821 committed:
[#114] Feat: add CompletedPlanCountByTimeChart UI

- Recharts 라이브러리의 Bar Chart와 Cartesian Grid를 사용하여 운동 시간대 별 완료된 플랜 갯수 그래프를 구현함
- 이번 달의 운동 시간대 별 총 횟수를 비교하여 최댓값을 가진 운동 시간을 문구에 명시함

soralee2821 committed:
[#114] Feat: add getDaysSinceSpecificDate utilFunc

- 가입일로부터 현재까지의 일수를 계산하는 함수를 구현함

soralee2821 committed:
[#114] Feat: add ChartComponents to Summary page

- TotalCountByAllWorkoutsChart, MonthlyTotalCountByEachWorkoutChart, CompletedPlanCountByTimeChart 컴포넌트를 추가함
- loginAtom의 상태를 확인하여 로그인 상태가 아닐 때 로그인 페이지로 이동하는 로직을 추가함
- userDataAtom의 joinDate 값과 getDaysSinceSpecificDate 유틸 함수를 사용하여 가입일로부터 현재까지의 일 수를 계산하여 문구로 보여줌

soralee2821 committed:
[#114] Feat: add api func, api path, query key as to totalWorkoutCountData

soralee2821 committed:
[#114] Feat: mock response of api func as to totalWorkoutCount

soralee2821 committed:
[#114] Refactor: use api func to get totalWorkoutCountData instead using constant

soralee2821 committed:
[#114] Feat: add api func, api path, query key as to monthWorkoutCountData

soralee2821 committed:
[#114] Feat: mock response of api func as to monthWorkoutCount

soralee2821 committed:
[#114] Refactor: use api func to get monthWorkoutCountData instead using constant

- monthWorkoutCountData의 map 메서드 내 콜백함수의 param 이름이 길고 data가 중복되므로 data 접두사를 지움

soralee2821 committed:
[#114] Feat: add api func, api path, query key as to completedPlanCountData

soralee2821 committed:
[#114] Feat: mock response of api func as to completedPlanCount

soralee2821 committed:
[#114] Refactor: use api func to get completedPlanCountData instead using constant

- completedPlanCountData의 map 메서드 내 콜백함수의 param 이름이 길고 data가 중복되므로 data 접두사를 지움

soralee2821 committed:
[#114] Feat: add configProvider to apply same custom style

- 프로필 이미지의 드롭다운과 같은 커스텀 스타일을 적용하기 위해 antd 라이브러이의 ConfigProvider 컴포넌트로 감쌈

soralee2821 committed:
[#114] Rename: rename dashboard constant file to summary

soralee2821 committed:
[#114] Refactor: add return type to useCompletedPlanCount

- type assertion을 지우기 위해 함수 반환 타입을 선언해줌

soralee2821 committed:
[#114] Refactor: add return type to useMonthWorkoutCount

- type assertion을 지우기 위해 함수 반환 타입을 선언해줌

soralee2821 committed:
[#114] Refactor: add return type to useTotalWorkoutCount

- type assertion을 지우기 위해 함수 반환 타입을 선언해줌

jaypedia commented:
드롭다운 UI에 arrow 같은 아이콘이 없어서 처음에 드롭다운이 아닌 줄 알았어요! 
arrow icon 같은 것을 넣는 건 어떨까요? 이런 식으로요
![image](https://github.com/Pullanner/pullanner-web/assets/85419343/8c14f398-75ea-4766-9b10-aa6d70a2fbfa)

jaypedia reviewed (commented):
차트 라이브러리를 잘 활용하셨네요! 

다만, 드롭다운 UI에 arrow 같은 아이콘이 없어서 처음에 드롭다운이 아닌 줄 알았어요! 
arrow icon 같은 것을 넣는 건 어떨까요? 이런 식으로요
![image](https://github.com/Pullanner/pullanner-web/assets/85419343/8c14f398-75ea-4766-9b10-aa6d70a2fbfa)

soralee2821 committed:
[#114] Feat: add down arrow icon to MonthDropdown

- 텍스트와 downArrow 아이콘을 수평 정렬하기 위해 flex 박스로 감싸주고, gap을 8px 설정함
- MonthDropdown의 padding-left, padding-right를 10px로 padding-top, padding-bottom을 6px로 변경함

jaypedia reviewed (approved):
빠르게 리뷰 반영해주셔서 좋았습니다~! 고생하셨어요!
